### ISSUE TITLE
[Bug] Unstable Performance of SAC Trainer rllib from Ray 1.10.0 and 1.11.0, compared to Ray 1.1.0

### ISSUE THREAD

**Issue by WeiSHI7** [ray-project/ray#23588]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### Issue Severity

Medium: It contributes to significant difficulty to complete my task but I work arounds and get it resolved.

### What happened + What you expected to happen

Hi, I am using class SACTrainer from ray.rllib.agents.sac to solve a problem within an external environment based on class ExternalEnv from ray.rllib.env.external_env. I run exactly the same codes with the same config in different conda envs with different versions of Ray, but I got very unstable outcomes when using the new versions of Ray such as Ray 1.10.0 and Ray 1.11.0.

![Capture](https://user-images.githubusercontent.com/79168422/160834522-6d36cd96-b083-4838-93e9-671e7a9c78da.PNG) 
This figure is the mean reward of episodes monitored by tensorboard. I expected that all the curves should follow the same or a similar trend as the black and green curves. However, as shown in the figure, the black curve and green curve are the outcomes produced on Ray 1.1.0 (TensorFlow 1.15.0). The orange and red curves are from Ray 1.11.0, and we can see there are big drops of reward during the learning process. I also tried to upgrade TensorFlow versions in the Ray 1.11.0 environment, but the outcomes remain the same.

I am wondering if there are some modifications to the updated versions of Ray that might influence the learning process of the trainer. Or maybe the updated Ray requires some modifications on config that I ignored.

### Versions / Dependencies

The experiments are run on Linux, python3.7, Ray 1.1.0, Ray 1.11.0, and Ray 1.10.0. Here are the libs used in the experiment
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import logging
import os
import sys

import ray
from ray.rllib.env.external_env import ExternalEnv
import ray.rllib.agents.sac as sac
from ray.rllib.models.tf.tf_modelv2 import TFModelV2
from ray.rllib.utils.annotations import override
from ray.rllib.models.tf.misc import normc_initializer
import gym
from ray.tune.logger import pretty_print
from ray.tune.registry import register_env
from ray.rllib.utils.framework import try_import_tf

from ray.rllib.models import ModelV2, ModelCatalog
from gym import ObservationWrapper
from ray.rllib.models.preprocessors import Preprocessor, get_preprocessor
from gym import spaces
from gym.wrappers import FlattenObservation
import numpy as np

### Reproduction script
```

And here is the config for my SACTrainer:
```
config={
            "log_level": "INFO",
            "observation_filter": "MeanStdFilter",
            "explore": myExplore,
            "in_evaluation": not myExplore,
            "gamma": 0.1,  #
            # === Model ===
            # Use two Q-networks (instead of one) for action-value estimation.
            # Note: Each Q-network will have its own target network.
            "twin_q": True,
            # Use a e.g. conv2D state preprocessing network before concatenating the
            # resulting (feature) vector with the action input for the input to
            # the Q-networks.
            "use_state_preprocessor": False,
            # Model options for the Q network(s).
            "Q_model": {
                "fcnet_activation": "relu",
                "fcnet_hiddens": [128, 128],
            },
            # Model options for the policy function.
            "policy_model": {
                "fcnet_activation": "relu",
                "fcnet_hiddens": [128, 128],
            },
            # Unsquash actions to the upper and lower bounds of env's action space.
            # Ignored for discrete action spaces.
            "normalize_actions": True,
            # === Evaluation ===
            # Evaluate with epsilon=0 every `evaluation_interval` training iterations.
            # The evaluation stats will be reported under the "evaluation" metric key.
            # Note that evaluation is currently not parallelized, and that for Ape-X
            # metrics are already only reported for the lowest epsilon workers.
            "evaluation_interval": None,
            # Number of episodes to run per evaluation period.
            "evaluation_num_episodes": 10,
            # N-step Q learning
            "n_step": 1,
            # === Learning ===
            # Disable setting done=True at end of episode. This should be set to True
            # for infinite-horizon MDPs (e.g., many continuous control problems).
            "no_done_at_end": False,
            # Update the target by \tau * policy + (1-\tau) * target_policy.
            "tau": 0.002,
            # Initial value to use for the entropy weight alpha.
            "initial_alpha": 1.0,
            # Target entropy lower bound. If "auto", will be set to -|A| (e.g. -2.0 for
            # Discrete(2), -3.0 for Box(shape=(3,))).
            # This is the inverse of reward scale, and will be optimized automatically.
            "target_entropy": None,
            # Number of env steps to optimize for before returning.
            "timesteps_per_iteration": 50,  # Default 1000,

            # === Replay buffer ===
            # Size of the replay buffer. Note that if async_updates is set, then
            # each worker will have a replay buffer of this size.
            "buffer_size": int(1e6),  # Default 50000,
            # If True prioritized replay buffer will be used.
            "prioritized_replay": True,
            # Alpha parameter for prioritized replay buffer.
            "prioritized_replay_alpha": 0.6,
            # Beta parameter for sampling from prioritized replay buffer.
            "prioritized_replay_beta": 0.4,
            # Time steps over which the beta parameter is annealed.
            "prioritized_replay_beta_annealing_timesteps": 200000,  # Default 20000,
            # Final value of beta
            "final_prioritized_replay_beta": 0.99,  # Default 0.4,
            # Epsilon to add to the TD errors when updating priorities.
            "prioritized_replay_eps": 1e-6,
            # Whether to LZ4 compress observations
            "compress_observations": False,
            # If set, this will fix the ratio of replayed from a buffer and learned on
            # timesteps to sampled from an environment and stored in the replay buffer
            # timesteps. Otherwise, the replay will proceed at the native ratio
            # determined by (train_batch_size / rollout_fragment_length).
            "training_intensity": None,

            # === Optimization ===
            "optimization": {
                "actor_learning_rate": 3e-4,
                "critic_learning_rate": 3e-4,
                "entropy_learning_rate": 3e-4,
            },
            # Update the target network every `target_network_update_freq` steps.
            "target_network_update_freq": 200,  # Default 0,
            # If not None, clip gradients during optimization at this value
            "grad_clip": None,
            # How many steps of the model to sample before learning starts.
            "learning_starts": 1,
            # Update the replay buffer with this many samples at once. Note that this
            # setting applies per-worker if num_workers > 1.
            "rollout_fragment_length": 4,  # Default 1,
            # Size of a batched sampled from replay buffer for training. Note that
            # if async_updates is set, then each worker returns gradients for a
            # batch of this size.
            "train_batch_size": 200,  # Default 256,

            # === Parallelism ===
            # Number of workers for collecting samples with. This only makes sense
            # to increase if your environment is particularly slow to sample, or if
            # you're using the Async or Ape-X optimizers.
            "num_workers": 0,
            # Whether to compute priorities on workers.
            "worker_side_prioritization": False,
            # Prevent iterations from going lower than this time span
            "min_iter_time_s": 1,
            "output": 'dlOrchResults_sac',
            "output_max_file_size": 500000,
        }
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by gjoliver** [comment]
curious, what is the workaround you used?

**Comment by ArturNiederfahrenhorst** [comment]
Hi @WeiSHI7 , @gjoliver,

I tried digging into this, which was quite hard since I fail to deploy ray 1.1 on my M1.
But here are some things I noted:

- `prioritized_replay_beta_annealing_timesteps ` and `final_prioritized_replay_beta` has no influence, since related code was removed at some point prior to 1.1. That was my first guess to be the difference between the two.
- Can you confirm how your models look? There have been some changes concerning that.
- Other than that the execution plans are very much the same
- We just implemented the training iteration function for SAC, which resembles the execution plan. Could you try on master?
- Since you have the full setup, can your maybe single out the release where the break occurs? We can't reproduce the experiment sadly.
